### PR TITLE
Adding base_namespace option support to Grpc.Tools

### LINKED ISSUE #17672: Grpc.Tools doesn't support same filename in different folders
### What version of gRPC and what language are you using?
Grpc.Tools 1.17.1
 
### What operating system (Linux, Windows, …) and version?
Windows 10
 
### What runtime / compiler are you using (e.g. python version or version of gcc)
C# MSBuild
 
### What did you do?
One project with two protos having the same filename but in different folders.

```
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFramework>netstandard2.0</TargetFramework>
    <Protobuf_StandardImportsPath>$(Protobuf_StandardImportsPath);..\protos</Protobuf_StandardImportsPath>
  </PropertyGroup>
  <ItemGroup>
    <PackageReference Include="Grpc.Tools" Version="1.17.1" PrivateAssets="All" />
    <Protobuf Include="..\protos\foo.proto" Link="protos\foo.proto" />
    <Protobuf Include="..\protos\dir1\bar.proto" Link="protos\dir1\bar.proto" />
    <Protobuf Include="..\protos\dir2\bar.proto" Link="protos\dir2\bar.proto" />
  </ItemGroup>
</Project>
```
 
### What did you expect to see?
The code should compile, generating 3 cs files.
 
### What did you see instead?
 ```
MSB3105: The item "obj\Debug\netstandard2.0\Bar.cs" was specified more than once in the "Sources" parameter.  Duplicate items are not supported by the "Sources" parameter.
```
 
### Anything else we should know about your project / environment?
We have an extensive library of protos that get reused by different projects that also share some protos. Therefore a common root path is necessary. The name `config.proto` for example is used in a couple of directories.

Until now, I used the following command to generate the files:
```
%protoc_path%\protoc.exe --proto_path=%protoc_path%\..;..\protos --csharp_out=%output_path% --csharp_opt=base_namespace --grpc_out=%output_path% --plugin=protoc-gen-grpc=%protoc-grpc_path%\grpc_csharp_plugin.exe ..\protos\foo.proto ..\protos\dir1\bar.proto ..\protos\dir2\bar.proto
```

The option `base_namespace` placed the generated files in sub-directories and avoided the problem. This problem will still remains for grpc service files, since there is no option like this for `grpc_opt`. But that problem can't be circumvented so easy right now.

/cc @kkm000

### PR DESCRIPTION
**Pull request by Falco20019**
I added support in two ways and kept the old behavior as default to not break any users. It's also more efficient if `Protobuf.BaseNamespaceEnabled` is kept on `false` as it requires looking into the file content to find out the correct path. Additionally to the attribute, I added a property `Protobuf_BaseNamespaceEnabledDefault` to enable it globally (in case a lot of entries would have to be enabled).

Blocked by #22627
Fixes #17672

/CC: @jtattermusch @kkm000

### CODE REVIEW

**Comment by Falco20019** [comment]
We should make sure that #22628 is kept in documentation when this is merged as it's more performant and working for most uses where relative paths are used. This solution is working in all use cases like absolute use cases and therefore still needed.

**Comment by jtattermusch** [comment]
While I understand that there might be some use for the `base_namespace` option when generating from the .proto files manually (or with a script), I think using `base_namespace` together with Grpc.Tools base_namespace support as a solution for https://github.com/grpc/grpc/issues/17672 is an overkill.

When using Grpc.Tools, you actually don't care about the exact path of the .cs files under obj, as long as they are unique (= 2 helloworld.proto files in different places need to generate 2 different .cs files under obj/Config/). So choosing a solution where you first introduce an option that generates namespace name based on C#-specific name mangling rules, and then try to get Grpc.Tools to mimic that mangled path (which requires analyzing file content with a regex in a very fragile way) sounds like overcomplicating things. Instead we should come up with some simple way of making the generated .cs file paths under obj/Config unique (I still believe this should be doable when the right trick is used), rather than bending ourselves backward trying to fix this with base_namespace and adding a lot of new complexity only to come up with a solution that is actually quite fragile.

File: src/csharp/Grpc.Tools/GeneratorServices.cs [hunk 1]
```diff
@@ -61,71 +62,114 @@ protected bool GrpcOutputPossible(ITaskItem proto)
...
+
+            return namespace_suffix.Replace('.', Path.DirectorySeparatorChar);
+        }
+
+        private string GetFileNamespace(string filename)
+        {
+            string data = File.ReadAllText(filename);
+
+            // First try to match the namespace from file options.
+            Match match = s_namespaceRegex.Match(data);
```
Comment (jtattermusch) [inline src/csharp/Grpc.Tools/GeneratorServices.cs:133]: matching the regexes like this is very fragile and dirty. We cannot do it like this.
A simple counterexample:  if I include `option  csharp_namespace` (with two spaces) instead of `option csharp_namespace` the regex won't catch that and you'll silently generate the wrong path and things will break in a very hard to detect way. This problem is fundamental and you'll never be able to come up with a regex that catches all the interesting cases reliably.

Basically the issue is that you're trying to emulate the c# namespace mangling logic from the plugin's C++ code, (without having access to any of the protobuf primitives that could tell you values of some of the options you need) and that is simply not going to work.

**Comment by Falco20019** [comment]
Will be superseded by the approach from here: https://github.com/grpc/grpc/issues/17672#issuecomment-621152800
